### PR TITLE
Add linux/arm64 platforms to image builds

### LINKED ISSUE #28: Add arm64 images
When developing on an apple silicon mac, docker uses linux/arm64 by default.
Using the amd64 image is possible, but is a lot slower.

It'd be nice if the CI also publishes arm64 images. Locally the Dockerfiles do seems to work for arm64 (at least v2.x).

### PR DESCRIPTION
**Pull request by timower**
This should allow the toolchain to be easily used on apple silicon macs (or any other arm64 device).

Closes #28

### CODE REVIEW

File: scripts/build [hunk 1]
```diff
@@ -48,6 +48,7 @@ version="$2"
 # Enable BuildKit for better cache behavior
 # See <https://docs.docker.com/engine/reference/builder/#buildkit>
 export DOCKER_BUILDKIT=1
```
Comment (Eeems) [inline scripts/build]:
```suggestion
```

**Review by Eeems — CHANGES_REQUESTED** [review]
My understanding is that if you use buildx, `DOCKER_BUILDKIT=1` is no longer needed

**Comment by timower** [comment]
I'll try something like https://docs.docker.com/build/ci/github-actions/multi-platform/#distribute-build-across-multiple-runners, otherwise I guess this isn't feasible

**Comment by Eeems** [comment]
To avoid timeouts due to how long it takes, probably. It might also make sense to split out the images into their own jobs as well. I've done that for the images I have generating nightly: https://github.com/Eeems/images/blob/main/.github/workflows/build.yml

**Comment by trixmoe** [comment]
GitHub seems to be adding M1 runners as part of their "[Standard GitHub-hosted runners for Public repositories](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories)", but they are not (yet) listed on their "[actions/runner-images](https://github.com/actions/runner-images?tab=readme-ov-file#available-images)" repo. They have an [open, recently (in hours) approved, but not yet merged PR](https://github.com/actions/runner-images/pull/9179), so I'm guessing it's coming very soon!

The upside is that this is native arm64, which would most likely run faster than emulated arm64 on amd64! Thus, this could possibly solve the timeout issue.

The macOS images don't seem to have docker pre-installed, but there are reports of [users simply installing it through homebrew](https://github.com/actions/runner-images/issues/8104)
The building scripts _seem_ like they'd run on macOS just fine, as they're just bash with a bunch of docker commands.

**Comment by trixmoe** [comment]
I've tried to bodge something together to show it'd work, but I'm stuck on a QEMU `HV_UNSUPPORTED` issue when starting colima ("container runtimes on macOS"). The previous reports were on `macos-12`, while this is running on `macos-14`, which is a beta for GitHub.

I've opened [an issue on the colima repo](https://github.com/abiosoft/colima/issues/970), but I'm guessing it might be a bug on GitHub's side.

**Comment by trixmoe** [comment]
My optimism in regards to M1 runners was short-sighted: GitHub (and simply Apple) does not support (nested) virtualization on those runners, which results in the VM not being able to boot up. There's thus no way to really improve the earlier amd64 Linux runner solution('s run time).

I found out about this through [the issue I opened on the colima repo](https://github.com/abiosoft/colima/issues/970), and the [announcement PR on actions/runner-images](https://github.com/actions/runner-images/issues/9254#issuecomment-1917916016), which pointed at [the GitHub docs](https://docs.github.com/en/actions/using-github-hosted-runners/about-larger-runners/about-larger-runners#limitations-for-macos-larger-runners):
> (...) nested-virtualization is not supported by arm64 runners

## Tl;dr of this whole ordeal
Stick with amd64 Linux runners (for the foreseeable future).
GH doesn't seem to be currently working on Linux arm64 runners (https://github.com/actions/runner-images/issues/5631).

**Comment by jandubois** [comment]
> GitHub (and simply Apple) does not support (nested) virtualization

Note that while GitHub claims it to be a limitation of Apple's Virtualization Framework, it really is a limitation of the M1 chip architecture; nested virtualization is supported by Apple's Virtualization Framework on M2 and M3 chips, which are based on a newer ARM spec.

**Comment by trixmoe** [comment]
> To avoid timeouts due to how long it takes, probably. It might also make sense to split out the images into their own jobs as well. I've done that for the images I have generating nightly: https://github.com/Eeems/images/blob/main/.github/workflows/build.yml

I've looked at the dependencies between the images, and I've realized that one could have 3 different stages, with each image having it's own separate job. Using different jobs also allows it to go over the 6 hour limit.

**Stage 1** would be the initial one, **Stage 2** starts once `toolchain` is done, **Stage 3** starts once `base` is done. Each of the blocks are separate jobs.

multi stage building

One would have to look into the exporting of images, to transfer them between jobs, for builds that are not published.
One could also separate amd64 and arm64 builds, but I'm unsure if this is doable when publishing to the image repository.

**Comment by Eeems** [comment]
I would recommend moving dotnet6 to stage 3 as well, make that dynamic, and have toolchain and base be static steps.

**Comment by trixmoe** [comment]
I looked more into multi-platform Docker images, as I had some time.

Separating the building of the images per platform is possible (exporting/importing single-platform images is no issue), but combining the images at the end is slightly hacky, as Docker does not support direct publishing of multi-platform "images" (manifest), but rather forces the individual images to be published already before creating the multi-platform manifest. (https://github.com/docker/cli/issues/3350)

To get around this, one can publish the single-platform images onto a [local registry ran through GHA services](https://docs.docker.com/build/ci/github-actions/local-registry/). Once they are published, create the manifest, and push it locally. Finally, it's possible to copy the multi-platform "image" to the public registry. (Similarly to https://github.com/rapidsai/gha-tools/pull/51)
